src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -544,24 +557,132 @@ void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo
+				MarkCustomAttribute (ca, reason);
+			} else {
+				// Record the custom attribute so that it has a reason, without actually marking it.
+				Tracer.AddDirectDependency (ca, reason, marked: false);
+			}
+
+			return true;
+		}
+
+		void MarkDynamicDependency (DynamicDependency dynamicDependency, MemberReference context)
+		{
+			Debug.Assert (context is MethodDefinition || context is FieldDefinition);
+			AssemblyDefinition assembly;
+			if (dynamicDependency.AssemblyName is string assemblyName) {
+				assembly = _context.GetLoadedAssembly (assemblyName);
+				if (assembly == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved assembly '{assemblyName}' in DynamicDependencyAttribute on '{context}'", 2031));
+					return;
 				}
+			} else {
+				assembly = context.Module.Assembly;
+				Debug.Assert (assembly != null);
+			}
 
-				return true;
+			TypeDefinition type;
+			if (dynamicDependency.TypeName is string typeName) {
+				type = assembly.FindTypeByDocumentationSignature (typeName);
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{typeName}' in DynamicDependencyAttribute on '{context}'", 2032));
+					return;
+				}
+			} else if (dynamicDependency.Type is TypeReference typeReference) {
+				type = typeReference.Resolve ();
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{typeReference}' in DynamicDependencyAtribute on '{context}'", 2032));
+					return;
+				}
+			} else {
+				type = context.DeclaringType.Resolve ();
+				if (type == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved type '{context.DeclaringType}' in DynamicDependencyAttribute on '{context}'", 2032));
+					return;
+				}
 			}
 
-			return false;
+			IEnumerable<IMemberDefinition> members;
+			if (dynamicDependency.MemberSignature is string memberSignature) {
+				members = type.FindMembersByDocumentationSignature (memberSignature);
+				if (!members.Any ()) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved member '{memberSignature}' in DynamicDependencyAttribute on '{context}'", 2033));
+					return;
+				}
+			} else {
+				var memberTypes = dynamicDependency.MemberTypes;
+				members = DynamicallyAccessedMembersBinder.GetDynamicallyAccessedMembers (type, memberTypes);
+				if (!members.Any ()) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"No members were resolved for '{memberTypes}' in DynamicDependencyAttribute on '{context}'", 2034));
+					return;
+				}
+			}
+
+			MarkMembers (type, members, new DependencyInfo (DependencyKind.DynamicDependency, dynamicDependency.OriginalAttribute));
+		}
+
+		void MarkMembers (TypeDefinition typeDefinition, IEnumerable<IMemberDefinition> members, DependencyInfo reason)
+		{
+			foreach (var member in members) {
+				if (member == typeDefinition) {
```

Comment:
Hmm - maybe we should have a "sentinel" value for this. Could be `null` or could be an artificial member (just new up empty member from Cecil). Not sure - I definitely don't like returning the type itself again - it lends itself to easy endless recursion bugs.

Reply:
I changed it to use a `null` sentinel.